flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/UserDefinedFunctionHelper.java

```diff
@@ -243,6 +244,36 @@ public static void prepareInstance(ReadableConfig config, UserDefinedFunction fu
         cleanFunction(config, function);
     }
 
+    /**
+     * Returns whether a {@link UserDefinedFunction} can be easily serialized and identified by only
+     * a fully qualified class name. It must have a default constructor and no serializable fields.
+     */
+    public static boolean isClassNameSerializable(UserDefinedFunction function) {
+        final Class<?> functionClass = function.getClass();
+        if (!InstantiationUtil.hasPublicNullaryConstructor(functionClass)) {
+            // function must be parameterized
+            return false;
+        }
+        Class<?> currentClass = functionClass;
+        while (!currentClass.equals(UserDefinedFunction.class)) {
+            for (Field field : currentClass.getDeclaredFields()) {
+                if (!Modifier.isTransient(field.getModifiers())
+                        && !Modifier.isStatic(field.getModifiers())) {
+                    // function seems to be stateful
+                    return false;
+                }
+            }
+            currentClass = currentClass.getSuperclass();
+        }
```

Comment:
That sounds like a very magic way to find out whether a function is stateful or not. Don't we have a marker interface for that?

Reply:
Even with a marker interface you need perform this check. User-defined function might come in various flavors. And the more interfaces, the harder it gets to get the job done.

Reply:
It is, but on the other hand this is very magic and very hard as a behavior to predict for the users. I would rather have this kind of logic enabled behind a flag and default to "if you wanna implement a stateful function, mark it as such with this interface"

Reply:
I'm also against magic but in this case the magic is minimal. It only influences a plan digest and makes it more deterministic. Most users don't parameterize their functions anyway and/or register it under a name which makes them non-anonymous anyway.